### ISSUE TITLE
mruby: how to get response body?

### ISSUE THREAD

**Issue by palxex** [nghttp2/nghttp2#751]
I'm to implement a debugging purpose proxy server with nghttpx + mruby, which need get the original response body and modify it. But in the [document](https://nghttp2.org/documentation/nghttpx.1.html#Nghttpx::Response) I found that Nghttp::Response only has headers and no body (in my imagination it should be a string). Is it possible to add it? Just made a try and found myself lost in memchunks.
EDIT: also Nghttpx::Request::body.

**Comment by tatsuhiro-t** [comment]
Currently, mruby hook is called only for headers phase.
I think it is not so hard to call mruby function when we got chunk of body.

**Comment by palxex** [comment]
Glad to see it available. Tried again and failed again, after move the invocation I can get body - but only part of it, when transfer-encoding is chunked. And after the move the env.resp.return cannot work as expected( it appends all heads after the string I send, don't know how to prevent it since not familiar with the state machine ). Too hard for a newbie:) Wish it available in next milestone.
A test point just thought: env.resp.body should always be a string, even if server response is gzipped.

**Comment by tatsuhiro-t** [comment]
I'm not sure how you did it, but you can access data with and without transfer-encoding.
When you access data, nghttpx has already committed headers to the remote endpoint, which cannot be undone, and env.resp.return cannot be used.
The downside of calling mruby function each time we get chunk of data is performance degradation, since mruby function invocation needs relatively high CPU cost.
I think mruby may have mrbgem to unzip data.  In general, nghttpx does not touch the content of proxied data.

**Comment by palxex** [comment]
> I'm not sure how you did it, but you can access data with and without transfer-encoding.

Let me explain what I did. I moved the mruby invocation from `HttpsUpstream::on_downstream_header_complete` to `HttpsUpstream::on_downstream_body_complete`, and add the following bridge code to shrpx_mruby_module_response.cc: 
```
namespace {
mrb_value response_get_body(mrb_state *mrb, mrb_value self) {
  auto data = static_cast<MRubyAssocData *>(mrb->ud);
  auto downstream = data->downstream;
  auto body_buf = downstream->get_response_buf();
  char *body;
  size_t bodylen = 0;

  body = (char *)malloc(body_buf->len);
  char *body_ptr = body;
  memset(body,0,body_buf->len);
  auto head = body_buf->head;
  int i=0;
  while(head){
    size_t chunk_len = head->len();
    std::copy_n(head->buf.begin(),chunk_len,body_ptr);
    body_ptr += chunk_len;
    bodylen += chunk_len;
    head = head->next;
    i++;
  }

  auto ai = mrb_gc_arena_save(mrb);

  auto body_t = mrb_str_new(mrb, body, bodylen);
  free(body);

  mrb_gc_arena_restore(mrb, ai);

  return body_t;
}
} // namespace
```
So it seems this place is just called last chunk? Could you give me a place to hook on every chunk arrives? 
> When you access data, nghttpx has already committed headers to the remote endpoint, which cannot be undone, and env.resp.return cannot be used.

Great...This explanation what I experienced. Maybe we'd introduce a new flag in on_req, when user set it on, all "immediately send to remote endpoint" behaviour could be delayed until on_resp is done? This will also makes other users who don't want it not feeling any performance downgrade. 
> The downside of calling mruby function each time we get chunk of data is performance degradation, since mruby function invocation needs relatively high CPU cost.

For those who need such a function, I think it's worth to pay some performance cost:) And for other people do not need it, they can just ignore the flag mentioned above and get performance as before. 
> I think mruby may have mrbgem to unzip data. In general, nghttpx does not touch the content of proxied data.

Maybe. But lets see the user experience. If the body is introduced, its mainly usage is MITM. For this usage, its' best just get the body string, modify it, and send it back. If it dont appear as a whole string, the whole function will be hard( if not completely unable) to work. Imagination there is simple task, process a long GET body, and modify all "orig" to "dest", for the current flow modal, it simple cannot be done, since the "orig" may break just on chunk edge. I suggest only introduce C hooks on chunk arrival, to cache data, unzip them( in fact I not saw a mrbgems handles it...but nghttp2 has a built-in inflator ) and concat to string, makes only one final mruby on_resp( maybe another callback is better? ) invocation. Then when user returns a modified string ,any transfer-encoding chunk header could be eliminated, content-encoding zip also, we just need calculate a new content-length and send them to the remote endpoint.

**Comment by tatsuhiro-t** [comment]
You can access response body in Upstream::on_downstream_body().

To access all response body in one string, we have to add bunch of new code to nghttpx, like storing response body, postponing response, etc, which makes nghttpx even more complicate piece of software.
Since this is not the primal use case of L7 Load Balancer, it is pretty much in low priority right now.

**Comment by palxex** [comment]
Agree. When start the thread I didn't imagine so much part related. I've
finished my requirement by utilizing mitmproxy, thanks for the discussion.

发自我的 iPod

在 2016-12-23，下午9:12，Tatsuhiro Tsujikawa <notifications@github.com> 写道：

You can access response body in Upstream::on_downstream_body().

To access all response body in one string, we have to add bunch of new code
to nghttpx, like storing response body, postponing response, etc, which
makes nghttpx even more complicate piece of software.
Since this is not the primal use case of L7 Load Balancer, it is pretty
much in low priority right now.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/nghttp2/nghttp2/issues/751#issuecomment-268987933>, or mute
the thread
